OpenSim/Common/Delay.h

```diff
@@ -0,0 +1,179 @@
+ * };
+ * @endcode
+ *
+ * This class is implemented via a SimTK::Measure_<T>::Delay.
+ *
+ * @ingroup operators
+ *
+ */
+template<typename T>
+class Delay : public Component {
+OpenSim_DECLARE_CONCRETE_OBJECT_T(Delay, T, Component);
+public:
+
+    /** @name Property declarations
+    These are the serializable properties associated with a Delay. **/
+    /**@{**/
+    OpenSim_DECLARE_PROPERTY(delay, double,
+        "The duration (nonnegative, seconds) by which the output is delayed.");
+    /**@}**/
+
+    /// Default constructor.
+    Delay();
+
+    /// Convenience constructor that sets the delay property.
+    explicit Delay(double delay);
+
+    /// Get the delayed value (the input's value at time t-delay).
+    T getValue(const SimTK::State& s) const;
+
+private:
+
+    void constructProperties() override;
+    void constructInputs() override;
+    void constructOutputs() override;
+
+    void extendFinalizeFromProperties() override;
+    void extendAddToSystem(SimTK::MultibodySystem& system) const override;
+
+    SimTK::MeasureIndex _delayMeasureIndex;
+};
+
+template<class T>
+Delay<T>::Delay() {
+    constructInfrastructure();
+}
+
+template<class T>
+Delay<T>::Delay(double delay) : Delay() {
+    set_delay(delay);
+}
+
+template<class T>
+void Delay<T>::constructProperties() {
+    constructProperty_delay(0.0);
+}
+
+template<class T>
+void Delay<T>::constructInputs() {
+    // TODO the requiredAt Stage should be the same as this Delay's output dependsOn stage.
+    constructInput<T>("input", SimTK::Stage::Time);
+}
+
+template<class T>
+void Delay<T>::constructOutputs() {
+    // TODO the depensdOn stage should be the dependsOn stage of the output
+    // that is wired to this Delay's input.
+    constructOutput<T>("output", &Delay::getValue, SimTK::Stage::Time);
+}
+
+template<class T>
+T Delay<T>::getValue(const SimTK::State& s) const {
+    const auto& subsys = getSystem().getDefaultSubsystem();
+    const auto& measure = subsys.getMeasure(_delayMeasureIndex);
+    return SimTK::Measure_<T>::Delay::getAs(measure).getValue(s);
+}
+
+template<class T>
+void Delay<T>::extendFinalizeFromProperties() {
+    Super::extendFinalizeFromProperties();
+    SimTK_VALUECHECK_NONNEG_ALWAYS(get_delay(),
+            "delay", "Delay::extendFinalizeFromProperties()");
+}
+
+template<class T>
+void Delay<T>::extendAddToSystem(SimTK::MultibodySystem& system) const {
+    Super::extendAddToSystem(system);
+    auto& sub = system.updDefaultSubsystem();
+    const auto& input = *static_cast<const Input<T>*>(&getInput("input"));
+    typename SimTK::Measure_<T>::Delay delayMeasure(sub,
+                                                    InputMeasure<T>(sub, input),
+                                                    get_delay());
+    const_cast<Delay*>(this)->_delayMeasureIndex =
+        delayMeasure.getSubsystemMeasureIndex();
+}
+
+typedef Delay<SimTK::Real> ScalarDelay;
+typedef Delay<SimTK::Vector> VectorDelay;
```

Comment:
I didn't know about default template arguments! What about in SWIG, though?

> Do the typedefs help SWIG or something?

Yes, the reason is because these can be the type names exposed in SWIG. Also, I think it's helpful to people who are not too comfortable with C++ templates. Is this case different from the existence of typedefs like Matrix2f or Vec3?